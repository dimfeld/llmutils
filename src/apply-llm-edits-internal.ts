import { $ } from 'bun';
import { processRawFiles } from './whole-file/parse_raw_edits.ts';
import { processXmlContents } from './xml/parse_xml.ts';
import { processSearchReplace } from './diff-editor/parse.ts';

export interface ApplyLlmEditsOptions {
  content: string;
  writeRoot?: string;
  dryRun?: boolean;
}

export interface ProcessFileOptions {
  content: string;
  writeRoot: string;
  dryRun?: boolean;
}

export async function applyLlmEdits({ content, writeRoot, dryRun }: ApplyLlmEditsOptions) {
  writeRoot ??= await getWriteRoot();
  const xmlMode = content.includes('<code_changes>');
  const diffMode = content.includes('<<<<<<< SEARCH');

  if (diffMode) {
    return await processSearchReplace({
      content,
      writeRoot,
      dryRun,
    });
  } else if (xmlMode) {
    return await processXmlContents({
      content,
      writeRoot,
      dryRun,
    });
  } else {
    return await processRawFiles({
      content,
      writeRoot,
      dryRun,
    });
  }
}

export async function getWriteRoot(cwd?: string) {
  return cwd || (await $`git rev-parse --show-toplevel`.text()).trim() || process.cwd();
}
