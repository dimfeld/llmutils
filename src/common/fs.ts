import * as fs from 'node:fs/promises';
import * as path from 'node:path';

import { debugLog } from '../logging.js';

/**
 * Function to perform the path validation for secure file operations
 */
export function validatePath(baseDir: string, relativePath: string): string {
  const absoluteBaseDir = path.resolve(baseDir);
  const absoluteTargetPath = path.resolve(absoluteBaseDir, relativePath);

  // Normalize paths to handle '.' and '..' components and different OS separators
  const normalizedBaseDir = path.normalize(absoluteBaseDir);
  const normalizedTargetPath = path.normalize(absoluteTargetPath);

  // Check if the normalized target path is within the normalized base directory
  if (
    !normalizedTargetPath.startsWith(normalizedBaseDir + path.sep) &&
    normalizedTargetPath !== normalizedBaseDir
  ) {
    throw new Error(
      `Security Error: Attempted file operation outside of the base directory "${normalizedBaseDir}". Target: "${normalizedTargetPath}"`
    );
  }
  return normalizedTargetPath;
}

export async function secureWrite(
  baseDir: string,
  relativePath: string,
  content: string | Buffer
): Promise<void> {
  const absoluteTargetPath = validatePath(baseDir, relativePath);

  debugLog(`Securely writing to: ${absoluteTargetPath}`);
  await Bun.write(absoluteTargetPath, content);
}

export async function secureRm(baseDir: string, relativePath: string): Promise<void> {
  const absoluteTargetPath = validatePath(baseDir, relativePath);

  debugLog(`Securely removing: ${absoluteTargetPath}`);
  // Use force: true to mimic `rm -f`, avoiding errors if the file doesn't exist.
  // Keep recursive: false as we typically expect to remove files, not directories here.
  await fs.rm(absoluteTargetPath, { force: true, recursive: false });
}