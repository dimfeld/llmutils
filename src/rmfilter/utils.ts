import { $ } from 'bun';
import type { SpawnOptions } from 'bun';

export let debug = false;
export let quiet = false;

export function setQuiet(value: boolean) {
  quiet = value;
}

export function setDebug(value: boolean) {
  debug = value;
}

// Helper function to log and execute commands
export function logSpawn<
  T extends SpawnOptions.OptionsObject<
    SpawnOptions.Writable,
    SpawnOptions.Readable,
    SpawnOptions.Readable
  >,
>(cmd: string[], options?: T) {
  if (debug) {
    console.log(`[DEBUG] Executing: ${cmd.join(' ')}`);
    if (options?.cwd) {
      console.log(`[DEBUG] cwd: ${options.cwd}`);
    }
  }

  if (quiet) {
    let opts = options || ({} as T);
    if (opts.stdout !== 'pipe') {
      opts.stdout = 'ignore';
    }

    if (opts.stderr !== 'pipe') {
      opts.stderr = 'ignore';
    }

    if (opts.stdio) {
      if (opts.stdio[0] !== 'pipe') {
        opts.stdio[0] = 'ignore';
      }

      if (opts.stdio[1] !== 'pipe') {
        opts.stdio[1] = 'ignore';
      }
    }
  }

  return Bun.spawn(cmd, options);
}

export type MaybeAwaited<T extends Promise<any>> = Awaited<T> | T;
export async function cachePromise<T extends Promise<any>>(
  cache: Map<string, MaybeAwaited<T>>,
  key: string,
  fn: () => T
): Promise<T> {
  let p = cache.get(key);
  if (p) return p;

  let result = await fn();
  cache.set(key, result);
  return result;
}

export type FnCache<T extends (...args: any[]) => any> = Map<string, MaybeAwaited<ReturnType<T>>>;

let cachedGitRoot: string | undefined;
export async function getGitRoot() {
  if (cachedGitRoot) {
    return cachedGitRoot;
  }

  const value = (await $`git rev-parse --show-toplevel`.nothrow().text()).trim() || process.cwd();

  cachedGitRoot = value;
  return value;
}
