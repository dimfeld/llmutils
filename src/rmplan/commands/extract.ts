// Command handler for 'rmplan extract'
// Converts a Markdown project plan into YAML

import * as path from 'path';
import chalk from 'chalk';
import * as clipboard from '../../common/clipboard.ts';
import { error, log } from '../../logging.js';
import { loadEffectiveConfig } from '../configLoader.js';
import { readPlanFile } from '../plans.js';
import { extractMarkdownToYaml, type ExtractMarkdownToYamlOptions } from '../process_markdown.js';
import { setQuiet } from '../../common/process.ts';
import type { PlanSchema } from '../planSchema.js';

export async function handleExtractCommand(inputFile: string | undefined, options: any) {
  setQuiet(options.quiet);

  let inputText: string;
  if (inputFile) {
    inputText = await Bun.file(inputFile).text();
  } else if (!process.stdin.isTTY) {
    inputText = await Bun.stdin.text();
  } else {
    inputText = await clipboard.read();
  }

  let outputPath = options.output;
  if (options.plan && !options.output) {
    let name = options.plan.endsWith('.yml')
      ? options.plan
      : path.basename(options.plan, '.md') + '.yml';
    outputPath = path.join(path.dirname(options.plan), name);
  }

  // Determine output path
  if (!outputPath) {
    throw new Error('Either --output or --plan must be specified');
  }

  const config = await loadEffectiveConfig(options.config);

  // Check if output file already exists and is a stub plan
  let stubPlanData: PlanSchema | undefined;
  const outputYmlPath = outputPath.endsWith('.yml') ? outputPath : `${outputPath}.yml`;
  try {
    const existingPlan = await readPlanFile(outputYmlPath);
    // Check if it's a stub plan (has structure but no tasks)
    if (existingPlan && (!existingPlan.tasks || existingPlan.tasks.length === 0)) {
      stubPlanData = existingPlan;
      if (!options.quiet) {
        log(chalk.blue('Found existing stub plan, preserving metadata'));
      }
    }
  } catch {
    // File doesn't exist or isn't valid YAML, that's fine
  }

  // Extract markdown to YAML using LLM
  const extractOptions: ExtractMarkdownToYamlOptions = {
    output: outputPath,
    projectId: options.projectId,
    issueUrls: options.issue ? [options.issue] : [],
    stubPlan: stubPlanData ? { data: stubPlanData, path: outputYmlPath } : undefined,
  };

  await extractMarkdownToYaml(inputText, config, options.quiet ?? false, extractOptions);
}
