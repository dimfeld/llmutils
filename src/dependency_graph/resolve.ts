import * as fs from 'fs/promises';
import * as path from 'path';
import * as YAML from 'js-yaml';
import {
  resolve as resolveExports,
  legacy as resolveLegacy,
  type Package as PackageJson,
} from 'resolve.exports';
import { packageUp } from 'package-up';
import { findUp } from 'find-up';
import { cachePromise, type FnCache } from '../rmfilter/utils.ts';

interface Package {
  name: string;
  path: string;
  packageJson: PackageJson;
  type: 'builtin' | 'dependency' | 'workspace';
}

export class Resolver {
  cachedImportPath: FnCache<typeof this.internalResolveImportPaths> = new Map();
  cachedRelativeImport: FnCache<typeof this.internalResolveRelativeImport> = new Map();
  cachedPackageImport: FnCache<typeof this.internalResolvePackageImport> = new Map();
  packageFromDir: FnCache<() => Promise<Package>> = new Map();

  packages: Map<string, Package> = new Map();
  pnpmWorkspacePath: string | undefined;
  resolveNodeModules = false;

  constructor(packages: Map<string, Package>) {
    this.packages = packages;
  }

  static async new(baseDir: string = process.cwd()) {
    let workspace = await Resolver.resolvePnpmWorkspace(baseDir);
    const resolver = new Resolver(workspace?.packages ?? new Map());
    resolver.pnpmWorkspacePath = workspace?.path;
    return resolver;
  }

  /**
   * Resolves import paths for a TypeScript file
   * @param filePath Path to the TypeScript file
   * @param imports Array of import specifiers from the file
   * @returns Map of import specifiers to resolved file paths
   */
  async resolveImportPaths(
    filePath: string,
    imports: string[]
  ): Promise<{ importPath: string; resolved: string | null }[]> {
    let baseDir = path.dirname(filePath);
    return Promise.all(
      imports.map((importPath) =>
        cachePromise(this.cachedImportPath, `${baseDir}:${importPath}`, async () => {
          let result = await this.internalResolveImportPaths(baseDir, importPath);
          if (result.resolved) {
            let tryPaths = [result.resolved.replace('/dist/', '/src/')];

            if (path.extname(tryPaths[0]) === '.js') {
              tryPaths.push(tryPaths[0].slice(0, -3) + '.ts');
            }

            let exists = await Promise.all(
              tryPaths.map(async (tryPath) => {
                try {
                  let exists = await Bun.file(tryPath).exists();
                  return exists ? tryPath : undefined;
                } catch {
                  return undefined;
                }
              })
            );

            result.resolved = exists.find((e) => e != null) ?? result.resolved;
          }

          return result;
        })
      )
    );
  }

  private async internalResolveImportPaths(
    baseDir: string,
    importSpecifier: string
  ): Promise<{ importPath: string; resolved: string | null }> {
    // Skip built-in Node.js modules
    if (importSpecifier.startsWith('node:')) {
      return { importPath: importSpecifier, resolved: null };
    }

    if (importSpecifier.startsWith('$lib/')) {
      let packagePath = await this.resolvePackageJson(baseDir);
      importSpecifier = path.join(packagePath.path, 'src', 'lib', importSpecifier.slice(5));
    }

    // Handle relative imports
    if (importSpecifier.startsWith('.') || importSpecifier.startsWith('/')) {
      const resolvedPath = await this.resolveRelativeImport(baseDir, importSpecifier);
      return { importPath: importSpecifier, resolved: resolvedPath };
    }

    // Handle package imports
    const resolvedPath = await this.resolvePackageImport(baseDir, importSpecifier);
    return { importPath: importSpecifier, resolved: resolvedPath };
  }

  /**
   * Resolves relative imports
   */
  async resolveRelativeImport(baseDir: string, importSpecifier: string): Promise<string | null> {
    return cachePromise(this.cachedRelativeImport, importSpecifier, () =>
      this.internalResolveRelativeImport(baseDir, importSpecifier)
    );
  }

  private async internalResolveRelativeImport(
    baseDir: string,
    importSpecifier: string
  ): Promise<string | null> {
    let resolvedPath = path.resolve(baseDir, importSpecifier);
    const extensions = ['.ts', '.tsx', '.js', '.jsx'];

    // Check if it's a directory with an index file
    if (
      await Bun.file(resolvedPath)
        .stat()
        .then((stats) => stats.isDirectory())
        .catch(() => false)
    ) {
      for (const ext of extensions) {
        const indexPath = path.join(resolvedPath, `index${ext}`);
        if (await Bun.file(indexPath).exists()) {
          return indexPath;
        }
      }
    }

    // Check with extensions
    const existingExt = path.extname(resolvedPath);
    let candidates: string[];
    if (!existingExt) {
      candidates = extensions.map((ext) => resolvedPath + ext);
    } else if (existingExt === '.js') {
      candidates = [resolvedPath, resolvedPath.replace('.js', '.ts')];
    } else if (existingExt === '.ts') {
      candidates = [resolvedPath, resolvedPath.replace('.ts', '.js')];
    } else if (existingExt === '.mjs') {
      candidates = [resolvedPath, resolvedPath.replace('.mjs', '.mts')];
    } else if (existingExt === '.mts') {
      candidates = [resolvedPath, resolvedPath.replace('.mts', '.mjs')];
    } else {
      candidates = [resolvedPath];
    }

    let exists = await Promise.all(
      candidates.map((candidate) =>
        Bun.file(candidate)
          .exists()
          .then((e) => (e ? candidate : undefined))
          .catch(() => undefined)
      )
    );

    return exists.find((candidate) => candidate != null) ?? null;
  }

  /**
   * Resolves package imports including export maps and pnpm workspaces
   */
  async resolvePackageImport(baseDir: string, importSpecifier: string): Promise<string | null> {
    return cachePromise(this.cachedPackageImport, `${baseDir}:${importSpecifier}`, () =>
      this.internalResolvePackageImport(baseDir, importSpecifier)
    );
  }

  private async internalResolvePackageImport(
    baseDir: string,
    importSpecifier: string
  ): Promise<string | null> {
    const thisPkg = await this.resolvePackageJson(baseDir);

    let subpathIndex = -1;
    if (importSpecifier.startsWith('@')) {
      // @namespace/name/subpath
      let firstSlash = importSpecifier.indexOf('/', 1);
      if (firstSlash > 0) {
        subpathIndex = importSpecifier.indexOf('/', firstSlash + 1);
      }
    } else {
      subpathIndex = importSpecifier.indexOf('/');
    }

    let depName = subpathIndex > 0 ? importSpecifier.substring(0, subpathIndex) : importSpecifier;
    let depSubpath = subpathIndex > 0 ? importSpecifier.slice(subpathIndex + 1) : '';

    let depVersion =
      thisPkg.packageJson.dependencies?.[depName] ?? thisPkg.packageJson.devDependencies?.[depName];

    let pkg: Package | undefined;
    if (depVersion?.startsWith('workspace:')) {
      pkg = this.packages.get(depName);
    } else if (this.resolveNodeModules) {
      const nodeModulesPath = path.join(thisPkg.path, 'node_modules', depName);
      const pkgJsonPath = path.join(nodeModulesPath, 'package.json');
      const pkgData = await Bun.file(pkgJsonPath).json();
      pkg = { name: depName, path: nodeModulesPath, packageJson: pkgData, type: 'builtin' };
    }

    if (!pkg) {
      return null;
    }

    // Check export maps using resolve.exports
    if (pkg.packageJson.exports) {
      const resolvedExport = resolveExports(pkg.packageJson, depSubpath || '.', {
        conditions: ['import', 'require', 'node', 'default'],
        unsafe: true, // Allow falling back to main/module fields
      });

      if (resolvedExport) {
        const exportPath = Array.isArray(resolvedExport) ? resolvedExport[0] : resolvedExport;
        return path.resolve(pkg.path, exportPath);
      }
    } else if (depSubpath) {
      return path.join(pkg.path, depSubpath);
    }

    const subpath =
      resolveLegacy(pkg.packageJson, {
        browser: false,
      }) || '';
    return path.join(pkg.path, subpath);
  }

  /** Find the package.json for a directory */
  async resolvePackageJson(dir: string): Promise<Package> {
    return cachePromise(this.packageFromDir, dir, async () => {
      const packageJsonPath = await packageUp({ cwd: dir });
      if (!packageJsonPath) {
        throw new Error('No package.json found in project');
      }

      const packageJson = await Bun.file(packageJsonPath).json();
      return {
        name: packageJson.name,
        path: path.dirname(packageJsonPath),
        packageJson,
        type: 'workspace',
      } satisfies Package;
    });
  }

  /**
   * Resolves pnpm workspace packages
   */
  private static async resolvePnpmWorkspace(baseDir: string): Promise<{
    packages: Map<string, Package>;
    path: string;
  } | null> {
    const pnpmWorkspaceYamlPath = await findUp('pnpm-workspace.yaml', {
      cwd: baseDir,
    });

    if (!pnpmWorkspaceYamlPath) return null;

    const content = await Bun.file(pnpmWorkspaceYamlPath).text();
    const workspacePatterns: string[] = (YAML.load(content) as any)?.packages;
    if (workspacePatterns?.length === 0) return null;

    const packageMap = new Map<string, Package>();
    const workspaceRoot = path.dirname(pnpmWorkspaceYamlPath);
    for (const pattern of workspacePatterns) {
      const workspaceDir = path.join(workspaceRoot, pattern.replace('/*', ''));
      const packages = await fs.readdir(workspaceDir).catch(() => []);
      for (const pkg of packages) {
        const pkgJsonPath = path.join(workspaceDir, pkg, 'package.json');
        try {
          const pkgData = await Bun.file(pkgJsonPath).json();

          packageMap.set(pkgData.name, {
            name: pkgData.name,
            path: path.join(workspaceDir, pkg),
            packageJson: pkgData,
            type: 'workspace',
          });
        } catch {
          continue;
        }
      }
    }

    return { packages: packageMap, path: pnpmWorkspaceYamlPath };
  }
}
