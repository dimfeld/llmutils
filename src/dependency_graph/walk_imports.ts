import { file } from 'bun';
import {
  extractExportedVariables,
  extractExportedFunctions,
  extractExportedClasses,
  extractImportsExportModules,
  extractExportedInterfaces,
  extractExportedTypeAliases,
  Extractor,
} from '../treesitter/extract.js';
import { Resolver } from './resolve.js';
import * as path from 'path';
import { error } from '../logging.js';

interface FileInfo {
  imports: { module: string; namedImports?: { name: string; alias?: string }[] }[];
  reexports: { module: string; namedExports?: { name: string; alias?: string }[] }[];
  exportedNames: Set<string>;
}

export class ImportWalker {
  private extractor: Extractor;
  private resolver: Resolver;
  private fileInfoCache: Map<string, FileInfo> = new Map();

  constructor(extractor: Extractor, resolver: Resolver) {
    this.extractor = extractor;
    this.resolver = resolver;
  }

  private async getFileInfo(filePath: string): Promise<FileInfo | null> {
    if (this.fileInfoCache.has(filePath)) {
      return this.fileInfoCache.get(filePath)!;
    }

    try {
      const code = await Bun.file(filePath).text();
      const language = path.extname(filePath) === '.svelte' ? 'svelte' : 'typescript';
      const parser = await this.extractor.createParser(language);
      let tree = parser.parse(code);
      if (!tree) {
        return null;
      }

      if (language === 'svelte') {
        let svelteTree = await this.extractor.getSvelteScript(tree);
        tree.delete();
        if (!svelteTree) {
          return null;
        }
        tree = svelteTree;
      }

      try {
        const { imports, reexports } = extractImportsExportModules(tree);
        const exportedNames = new Set<string>();

        // TODO replace all these with a single treesitter query and/or combine with extractImportsExportModules
        const variables = extractExportedVariables(tree);
        for (const v of variables) {
          exportedNames.add(v.name);
        }

        const functions = extractExportedFunctions(tree);
        for (const f of functions) {
          exportedNames.add(f.name);
        }

        const classes = extractExportedClasses(tree);
        for (const c of classes) {
          exportedNames.add(c.name);
        }

        const interfaces = extractExportedInterfaces(tree);
        for (const i of interfaces) {
          exportedNames.add(i.name);
        }

        const typeAliases = extractExportedTypeAliases(tree);
        for (const t of typeAliases) {
          exportedNames.add(t.name);
        }

        const fileInfo: FileInfo = {
          imports,
          reexports,
          exportedNames,
        };

        this.fileInfoCache.set(filePath, fileInfo);
        return fileInfo;
      } finally {
        tree.delete();
      }
    } catch (e) {
      error(`Error parsing file ${filePath}:`, e);
      return null;
    }
  }

  private async findDefiningFile(
    variableName: string,
    modulePath: string,
    visited: Set<string> = new Set()
  ): Promise<string | null> {
    if (visited.has(modulePath)) {
      return null; // Cycle detected
    }
    visited.add(modulePath);

    const fileInfo = await this.getFileInfo(modulePath);
    if (!fileInfo) {
      return null;
    }

    if (fileInfo.exportedNames.has(variableName)) {
      return modulePath;
    }

    for (const reexport of fileInfo.reexports) {
      if (reexport.namedExports) {
        for (const exp of reexport.namedExports) {
          if (exp.alias === variableName || (!exp.alias && exp.name === variableName)) {
            const originalName = exp.name;
            const resolved = await this.resolver.resolveImportPaths(modulePath, [reexport.module]);
            if (resolved[0].resolved) {
              const definingFile = await this.findDefiningFile(
                originalName,
                resolved[0].resolved,
                visited
              );
              if (definingFile) {
                return definingFile;
              }
            }
          }
        }
      } else {
        // `export * from 'another'`
        const resolved = await this.resolver.resolveImportPaths(modulePath, [reexport.module]);
        if (resolved[0].resolved) {
          const definingFile = await this.findDefiningFile(
            variableName,
            resolved[0].resolved,
            visited
          );
          if (definingFile) {
            return definingFile;
          }
        }
      }
    }

    return null;
  }

  async getDefiningFiles(filePath: string): Promise<Set<string>> {
    const fileInfo = await this.getFileInfo(filePath);
    if (!fileInfo) {
      return new Set();
    }

    const definingFiles = new Set<string>();

    const allResolved = await this.resolver.resolveImportPaths(
      filePath,
      fileInfo.imports.map((i) => i.module)
    );
    for (let i = 0; i < allResolved.length; i++) {
      let resolved = allResolved[i];
      if (!resolved.resolved) {
        continue;
      }
      let imp = fileInfo.imports[i];

      const modulePath = resolved.resolved;

      if (imp.namedImports) {
        for (const namedImport of imp.namedImports || []) {
          const variableName = namedImport.name;
          const definingFile = await this.findDefiningFile(variableName, modulePath);
          if (definingFile) {
            definingFiles.add(definingFile);
          }
        }
      } else {
        definingFiles.add(modulePath);
      }
    }

    return definingFiles;
  }

  /** Get the import tree of a file and all the files it imports. If processing multiple files at once,
   * you can define `seen` yourself and pass it in to avoid duplicate work. */
  async getImportTree(filePath: string, seen: Set<string> = new Set()): Promise<Set<string>> {
    if (seen.has(filePath)) {
      return seen;
    }

    let fileInfo = await this.getFileInfo(filePath);
    if (!fileInfo) {
      return seen;
    }

    seen.add(filePath);

    let resolved = await this.resolver.resolveImportPaths(filePath, [
      ...fileInfo.imports.map((imp) => imp.module),
      ...fileInfo.reexports.map((reexport) => reexport.module),
    ]);

    let modulePaths = resolved.map((r) => r.resolved).filter((r) => r != null);

    await Promise.all(modulePaths.map((modulePath) => this.getImportTree(modulePath, seen)));

    return seen;
  }
}
