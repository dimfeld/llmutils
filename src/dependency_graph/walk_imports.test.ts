import { expect, test, describe, afterAll, beforeAll } from 'bun:test';
import * as path from 'node:path';
import * as fs from 'fs/promises';
import * as os from 'os';

import { packageUp } from 'package-up';
import { ImportWalker } from './walk_imports';
import { Extractor } from '../treesitter/extract.ts';
import { Resolver } from './resolve.ts';

describe('getDefiningFiles', () => {
  test('walk imports from walk_imports.ts', async () => {
    const rootDir = path.dirname((await packageUp())!);
    const walker = new ImportWalker(new Extractor(), await Resolver.new(rootDir));

    const imports = await walker.getDefiningFiles(
      path.join(rootDir, 'src/dependency_graph/walk_imports.ts')
    );
    const relativeImports = imports
      .values()
      .map((f) => path.relative(rootDir, f))
      .toArray()
      .sort();
    expect(relativeImports).toMatchSnapshot();
  });

  test('walk imports from apply-llm-edits.ts', async () => {
    const rootDir = path.dirname((await packageUp())!);
    const walker = new ImportWalker(new Extractor(), await Resolver.new(rootDir));

    const imports = await walker.getDefiningFiles(
      path.join(rootDir, 'src/apply-llm-edits/apply.ts')
    );
    const relativeImports = imports
      .values()
      .map((f) => path.relative(rootDir, f))
      .toArray()
      .sort();
    expect(relativeImports).toMatchSnapshot();
  });
});

describe('findImporters', () => {
  let tempImporterTestDir: string;

  async function setupImporterTestStructure(): Promise<string> {
    const tempDir = await fs.mkdtemp(path.join(os.tmpdir(), 'importer-test-'));

    // pkg-a
    await fs.mkdir(path.join(tempDir, 'pkg-a'));
    await Bun.write(
      path.join(tempDir, 'pkg-a', 'package.json'),
      JSON.stringify({ name: 'pkg-a', version: '1.0.0' })
    );
    await Bun.write(
      path.join(tempDir, 'pkg-a', 'importer.ts'),
      `
    import { util } from './utils'; // Relative import
    import { feature } from 'pkg-b/feature'; // Package import via export map
    import { helper } from 'pkg-b/utils/helper'; // Wildcard export
    import { anotherUtil } from "./anotherDir/anotherUtil";

    console.log(util(), feature(), helper(), anotherUtil());
  `
    );
    await Bun.write(path.join(tempDir, 'pkg-a', 'utils.ts'), `export const util = () => 'util';`);
    await fs.mkdir(path.join(tempDir, 'pkg-a', 'anotherDir'));
    await Bun.write(
      path.join(tempDir, 'pkg-a', 'anotherDir', 'anotherUtil.ts'),
      `export const anotherUtil = () => 'another util';`
    );

    // pkg-b
    await fs.mkdir(path.join(tempDir, 'pkg-b'));
    await Bun.write(
      path.join(tempDir, 'pkg-b', 'package.json'),
      JSON.stringify({
        name: 'pkg-b',
        version: '1.0.0',
        exports: {
          '.': './main.ts', // Export for pkg-b itself
          './feature': './src/feature.ts', // Export for pkg-b/feature
          './utils/*': './src/utils/*.ts', // Wildcard export
        },
      })
    );
    await fs.mkdir(path.join(tempDir, 'pkg-b', 'src', 'utils'), {
      recursive: true,
    });
    await Bun.write(
      path.join(tempDir, 'pkg-b', 'src', 'utils', 'helper.ts'),
      `export const helper = () => 'helper';`
    );
    await Bun.write(
      path.join(tempDir, 'pkg-b', 'main.ts'),
      `export const mainFunc = () => 'main from b';`
    );
    await Bun.write(
      path.join(tempDir, 'pkg-b', 'src', 'feature.ts'),
      `export const feature = () => 'feature';`
    );
    // A file in pkg-b that imports its own feature.ts
    await Bun.write(
      path.join(tempDir, 'pkg-b', 'internal-consumer.ts'),
      `
    import { feature } from './src/feature'; // Direct relative import
    console.log(feature());
  `
    );

    // pkg-c (imports from pkg-b root)
    await fs.mkdir(path.join(tempDir, 'pkg-c'));
    await Bun.write(
      path.join(tempDir, 'pkg-c', 'package.json'),
      JSON.stringify({ name: 'pkg-c', version: '1.0.0' })
    );
    await Bun.write(
      path.join(tempDir, 'pkg-c', 'consumer.ts'),
      `
    import { mainFunc } from 'pkg-b';
    console.log(mainFunc());
  `
    );

    // pnpm-workspace.yaml
    await Bun.write(path.join(tempDir, 'pnpm-workspace.yaml'), "packages:\n  - '*'");

    return tempDir;
  }

  beforeAll(async () => {
    tempImporterTestDir = await setupImporterTestStructure();
  });

  afterAll(async () => {
    if (tempImporterTestDir) {
      await fs.rm(tempImporterTestDir, { recursive: true, force: true });
    }
  });

  test('findImporters for a local file (pkg-a/utils.ts)', async () => {
    const walker = new ImportWalker(new Extractor(), await Resolver.new(tempImporterTestDir));
    const targetFile = path.join(tempImporterTestDir, 'pkg-a', 'utils.ts');
    const importers = await walker.findImporters(targetFile);

    const relativeImporters = Array.from(importers)
      .map((f) => path.relative(tempImporterTestDir, f))
      .sort();
    expect(relativeImporters).toEqual([path.join('pkg-a', 'importer.ts')]);
  });

  test('findImporters for a local file in a subdirectory (pkg-a/anotherDir/anotherUtil.ts)', async () => {
    const walker = new ImportWalker(new Extractor(), await Resolver.new(tempImporterTestDir));
    const targetFile = path.join(tempImporterTestDir, 'pkg-a', 'anotherDir', 'anotherUtil.ts');
    const importers = await walker.findImporters(targetFile);

    const relativeImporters = Array.from(importers)
      .map((f) => path.relative(tempImporterTestDir, f))
      .sort();
    expect(relativeImporters).toEqual([path.join('pkg-a', 'importer.ts')]);
  });

  test('findImporters for an exported file in another package (pkg-b/src/feature.ts)', async () => {
    const walker = new ImportWalker(new Extractor(), await Resolver.new(tempImporterTestDir));
    // Target is the actual source file path
    const targetFile = path.join(tempImporterTestDir, 'pkg-b', 'src', 'feature.ts');
    const importers = await walker.findImporters(targetFile);

    const relativeImporters = Array.from(importers)
      .map((f) => path.relative(tempImporterTestDir, f))
      .sort();
    expect(relativeImporters).toEqual(
      [
        path.join('pkg-a', 'importer.ts'), // Imports 'pkg-b/feature'
        path.join('pkg-b', 'internal-consumer.ts'), // Imports './src/feature'
      ].sort()
    );
  });

  test('findImporters for a wildcard exported file (pkg-b/src/utils/helper.ts)', async () => {
    const walker = new ImportWalker(new Extractor(), await Resolver.new(tempImporterTestDir));
    const targetFile = path.join(tempImporterTestDir, 'pkg-b', 'src', 'utils', 'helper.ts');
    const importers = await walker.findImporters(targetFile);

    const relativeImporters = Array.from(importers)
      .map((f) => path.relative(tempImporterTestDir, f))
      .sort();
    expect(relativeImporters).toEqual([path.join('pkg-a', 'importer.ts')].sort());
  });

  test('findImporters for a package root export (pkg-b/main.ts)', async () => {
    const walker = new ImportWalker(new Extractor(), await Resolver.new(tempImporterTestDir));
    const targetFile = path.join(tempImporterTestDir, 'pkg-b', 'main.ts');
    const importers = await walker.findImporters(targetFile);

    const relativeImporters = Array.from(importers)
      .map((f) => path.relative(tempImporterTestDir, f))
      .sort();
    expect(relativeImporters).toEqual(
      [
        path.join('pkg-c', 'consumer.ts'), // Imports 'pkg-b'
      ].sort()
    );
  });

  test('findImporters for a reexported file (pkg-b/src/utils/helper.ts reexported)', async () => {
    // Create a reexport file in pkg-b
    await Bun.write(
      path.join(tempImporterTestDir, 'pkg-b', 'src', 'utils', 'reexport.ts'),
      `export { helper } from './helper';`
    );
    // Create a file in pkg-c that imports the reexport
    await Bun.write(
      path.join(tempImporterTestDir, 'pkg-c', 'reexport-consumer.ts'),
      `
    import { helper } from 'pkg-b/utils/reexport';
    console.log(helper());
  `
    );

    const walker = new ImportWalker(new Extractor(), await Resolver.new(tempImporterTestDir));
    const targetFile = path.join(tempImporterTestDir, 'pkg-b', 'src', 'utils', 'helper.ts');
    const importers = await walker.findImporters(targetFile);

    const relativeImporters = Array.from(importers)
      .map((f) => path.relative(tempImporterTestDir, f))
      .sort();
    expect(relativeImporters).toEqual(
      [
        path.join('pkg-a', 'importer.ts'), // Direct import
        path.join('pkg-c', 'reexport-consumer.ts'), // Via reexport
      ].sort()
    );
  });

  test('findImporters for a file with no importers', async () => {
    const lonelyFilePath = path.join(tempImporterTestDir, 'pkg-a', 'lonely.ts');
    await Bun.write(lonelyFilePath, 'export const lonely = true;');

    const walker = new ImportWalker(new Extractor(), await Resolver.new(tempImporterTestDir));
    const importers = await walker.findImporters(lonelyFilePath);

    expect(Array.from(importers)).toEqual([]);
  });
});
