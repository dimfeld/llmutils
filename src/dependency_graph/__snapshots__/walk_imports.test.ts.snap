// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`walk imports from walk_imports.ts 1`] = `
[
  "src/common/file_finder.ts",
  "src/dependency_graph/resolve.ts",
  "src/logging.ts",
  "src/treesitter/extract.ts",
]
`;

exports[`walk imports from apply-llm-edits.ts 1`] = `
[
  "src/apply-llm-edits/failures.ts",
  "src/apply-llm-edits/interactive.ts",
  "src/editor/diff-editor/parse.ts",
  "src/editor/types.ts",
  "src/editor/udiff-simple/parse.ts",
  "src/editor/whole-file/parse_raw_edits.ts",
  "src/editor/xml/parse_xml.ts",
  "src/logging.ts",
  "src/rmfilter/repomix.ts",
  "src/rmfilter/rmfilter.ts",
  "src/rmfilter/utils.ts",
]
`;

exports[`getDefiningFiles walk imports from walk_imports.ts 1`] = `
[
  "src/common/file_finder.ts",
  "src/dependency_graph/filenames.ts",
  "src/dependency_graph/resolve.ts",
  "src/logging.ts",
  "src/treesitter/extract.ts",
]
`;

exports[`getDefiningFiles walk imports from apply-llm-edits.ts 1`] = `
[
  "src/apply-llm-edits/failures.ts",
  "src/apply-llm-edits/interactive.ts",
  "src/editor/diff-editor/parse.ts",
  "src/editor/types.ts",
  "src/editor/udiff-simple/parse.ts",
  "src/editor/whole-file/parse_raw_edits.ts",
  "src/editor/xml/parse_xml.ts",
  "src/logging.ts",
  "src/rmfilter/repomix.ts",
  "src/rmfilter/rmfilter.ts",
  "src/rmfilter/utils.ts",
]
`;
